src/coreclr/vm/amd64/jithelpers_fast.S

```diff
@@ -91,63 +87,60 @@ LEAF_ENTRY JIT_ByRefWriteBarrier, _TEXT
     // Additionally we know for sure that we are inside the heap and therefore don't
     // need to replicate the above checks.
     LOCAL_LABEL(DoneShadow_ByRefWriteBarrier):
+        pop     rax
+        pop     r11
+        pop     r10
 #endif
 
 #ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
         // Update the write watch table if necessary
-        PREPARE_EXTERNAL_VAR g_sw_ww_enabled_for_gc_heap, rax
-        cmp     byte ptr [rax], 0x0
+        cmp     byte ptr [C_VAR(g_sw_ww_enabled_for_gc_heap)], 0x0
         je      LOCAL_LABEL(CheckCardTable_ByRefWriteBarrier)
         mov     rax, rdi
         shr     rax, 0xC // SoftwareWriteWatch::AddressToTableByteIndexShift
-        PREPARE_EXTERNAL_VAR g_sw_ww_table, r10
-        add     rax, qword ptr [r10]
+        add     rax, qword ptr [C_VAR(g_sw_ww_table)]
         cmp     byte ptr [rax], 0x0
         jne     LOCAL_LABEL(CheckCardTable_ByRefWriteBarrier)
         mov     byte ptr [rax], 0xFF
 #endif
 
-    LOCAL_LABEL(CheckCardTable_ByRefWriteBarrier):
         // See if we can just quick out
-        PREPARE_EXTERNAL_VAR g_ephemeral_low, rax
-        cmp     rcx, [rax]
+    LOCAL_LABEL(CheckCardTable_ByRefWriteBarrier):
+        cmp     rcx, [C_VAR(g_ephemeral_low)]
         jb      LOCAL_LABEL(Exit_ByRefWriteBarrier)
-        PREPARE_EXTERNAL_VAR g_ephemeral_high, rax
-        cmp     rcx, [rax]
+        cmp     rcx, [C_VAR(g_ephemeral_high)]
         jnb     LOCAL_LABEL(Exit_ByRefWriteBarrier)
 
+        // do the following checks only if we are allowed to trash rax
+        // otherwise we don't have enough registers
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
```

Comment:
This ifdef does not make sense. The code under the ifdef has nothing to do with software write watch.

I think you want to drop the comment that RAX is trashed only when `FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP` is defined. Instead, just say the RAX is trashed, end of story.